### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by MichaelRotaru**
1. Code refactoring
2. Rebuild function that trigger “Create subscription” to match requested parameters
3. Rebuild function that trigger “Create payment” service to match requested parameters
4. Added Automatic Recurring Payment functionality for subscriptions witch, is triggered by 2 cases:
    - After payment authorization
    - On due date
5. Added Cancel Subscription functionality, can be triggered by 2 cases:
    - Manual from Admin > Order screen
    - Automatic at expiration
6. Added Refund functionality for simple and recurring payments, can be triggered from Admin > Invoice screen
7. Tested 3D secure feature - in order to activate it the client should add “3d-” string before Payment Method in Quickpay Configuration Pannel. Ex. “creditcard” = non-3dsecure payment option, “3d-creditcard” = 3d secure payment option

### CODE REVIEW

File: quickpay.php [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+ * @param array $params Payment Gateway Module Parameters
+ *
+ * @return array Request parameters
+ */
+function quickpay_request_params($params)
+{
+    /** Request array */
+    $request_arr = [ 'currency' => $params['currency']
+                   , 'order_id' => sprintf('%s%04d%s', $params['prefix'], $params['invoiceid'], isset($params['suffix'])?$params['suffix']:'')
+                   , 'description' => $params['description']
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:439]: It is specified as a parameter just subscriptions: https://learn.quickpay.net/tech-talk/api/services/#POST-subscriptions---format-

Screenshot 2020-12-11 at 11 06 16

File: quickpay.php [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+                       ];
+
+    $request_arr['invoice_address'] = $invoice_address;
+
+    /** Extract the invoice items details. */
+    $invoice = localAPI(/**command*/'GetInvoice', /**postData*/['invoiceid' => $params['invoiceid']]);
+
+    $basket = [];
+    foreach ($invoice['items']['item'] as $item) {
+        $basket[] = [ 'qty' => 1
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:460]: I've tested with multiple products and if i choose multiple quantities one product instance is made for each item. I also seen in other gateways examples that is used 1 for quantity.

Screenshot 2020-12-11 at 11 13 19

File: quickpay.php [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+                       , 'company_name' => $params['clientdetails']['companyname']
+                       , 'street' => (!empty($params['clientdetails']['address2'])) ? ($params['clientdetails']['address1'] . ', ' . $params['clientdetails']['address2']) : ($params['clientdetails']['address1'])
+                       , 'city' => $params['clientdetails']['city']
+                       , 'zip_code' => $params['clientdetails']['postcode']
+                       , 'region' => $params['clientdetails']['state']
+                       , 'phone_number' => $params['clientdetails']['phonenumber']
+                       , 'email' => $params['clientdetails']['email']
+                       ];
+
+    $request_arr['invoice_address'] = $invoice_address;
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:453]: No shipping address by default, just billing.

File: quickpay.php [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+ * @return string - Payment URL | Null
+ * @throws Exception
+ */
+function quickpay_create_payment_link($payment, $params, $type = 'payment')
+{
+    /** Quickpay API key */
+    $apiKey = $params['apikey'];
+
+    /** Gateway request parameters array */
+    $request = [ "amount"                       => str_replace('.', '', $params['amount'])
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:225]: If we convert the string to integer we lose the decimals. That's how they did it in the previous version.

File: quickpay.php [hunk 5]
```diff
@@ -0,0 +1,664 @@
...
+/************************** Helper functions START **************************/
+
+/**
+ * Signs the setup parameters
+ * @param array $params
+ * @param array $api_key
+ *
+ * @return string encoded string
+ */
+function sign($params, $api_key)
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:621]: That was defined in the previous version, i'll remove it.

File: quickpay.php [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+      );
+    }
+
+    /** Success */
+    return array(
+        /** 'success' if successful, any other value for failure */
+        'status' => 'success',
+        /** Data to be recorded in the gateway log - can be a string or array */
+        'rawdata' => 'Subscription successfully canceled',
+    );
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:362]: Yes. Both logTransaction() and logActivity() are called.

File: quickpay.php [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+    error_log('Refund call - Test msg');
+
+    logTransaction(/**gatewayName*/'quickpay', /**debugData*/['params' => $params], __FUNCTION__ . '::' . 'Refund request received');
+
+    /** Get invoice data */
+    $invoice = localAPI(/**command*/'GetInvoice', /**postData*/['invoiceid' => $params['invoiceid']]);
+
+    /** Gateway request parameters array */
+    $request = [ 'id'       => $params['transid']
+               , 'amount'   => str_replace('.', '', $params['amount'])
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:387]: Same as above.

File: callback/quickpay.php [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+                    require_once __DIR__ . '/../../../modules/gateways/quickpay.php';
+                    /** Payment link from response */
+                    $linkArray = json_decode(json_encode($request->link), TRUE);
+                    /** Recurring payment parameters */
+                    $params = [ "amount"                        => number_format(($linkArray['amount']/100.0), 2, '.', '') /** Convert amount to decimal type */
+                              , "continue_url"                  => $linkArray['continue_url']
+                              , "cancel_url"                    => $linkArray['cancel_url']
+                              , "callback_url"                  => $linkArray['callback_url']
+                              , "customer_email"                => $linkArray['customer_email']
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply callback/quickpay.php:104]: Is alright. I've checked other gateways and they do the same.

File: quickpay.php [hunk 9]
```diff
@@ -0,0 +1,664 @@
...
+
+    /** Check if transaction type is recurring */
+    if ($type === 'recurring') {
+        /** Construt orderid string */
+        $request["order_id"] = sprintf('%s%04d_r', $orderPrefix, $params['invoiceid']);
+        /** Request endpoint */
+        $endpoint = sprintf('subscriptions/%s/recurring', $payment->id/** Subscription_id */);
+        $response = quickpay_request($apiKey, $endpoint, $request, 'POST');
+        /** Current tranzaction id */
+        $payment->id = $response->id;
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:252]: In order to match the simple create_payment_link flow where the function call parameter is an object.

Screenshot 2020-12-11 at 12 43 14

File: quickpay.php [hunk 10]
```diff
@@ -0,0 +1,664 @@
...
+                //TODO - switch to systemurl
+               , "callback_url"                 => 'http://a160c3b18df9.ngrok.io/whmcs/' . 'modules/gateways/callback/' . $params['paymentmethod'] . '.php'//$params['systemurl']
+               , "customer_email"               => $params['clientdetails']['email']
+               , "payment_methods"              => $params['payment_methods']
+               , "language"                     => $params['language']
+               , "auto_capture"                 => $params['autocapture']
+                /** TYPO FIX */
+               , "autocapture"                  => $params['autocapture']
+               , "autofee"                      => $params['autofee']
+                /** TYPO FIX */
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:237]: This parameter is required in both forms forms "autofee" and "auto_fee". I'll add a proper comment.

Screenshot 2020-12-11 at 11 35 09
Screenshot 2020-12-11 at 11 36 12

[the remark this answers is not in the document]
Comment (cosminbosutar) [reply quickpay.php:237]: So the approach is to add them both in order to make sure that you always satisfy both request types?

File: quickpay.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (cosminbosutar) [reply quickpay.php:225]: Can we convert to float/double?

What is the decimal separator character '.' or ','?
My concern is to not have the '.' character as decimal separator and to transform number "11.23" into "1123".

File: quickpay.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (cosminbosutar) [reply quickpay.php:252]: From what I've seen in the "create_payment_link" function, there is no other member besides "id" used. Therefore I assume there is no need to send an object, and the ID would be sufficient.

What am I missing?

File: callback/quickpay.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (cosminbosutar) [reply callback/quickpay.php:104]: I still don't understand how this works. 
From my perspective we call the "quickpay_create_payment_link" function with the "$params" variable defined inside the "callback/quickpay.php" file and inside the "quickpay_create_payment_link" function we try to extract from the "$params" argument the following "$params['clientdetails']['email']" that is not defined.

Can you please elaborate?

File: quickpay.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:237]: Yes, the one is not used will be ignored.

File: quickpay.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:225]: The amount parameter is type integer so the only way to send float values is la that (1234 in request = 11.23 in quickpay). The last 2 digits will be interpreted as decimals. Decimal separator is always "." . 
I can do this "number_format( $price * 100, 0, '', '' )" but is kinda same thing.

File: quickpay.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply quickpay.php:252]: Sorry my bad. I misunderstood.

File: callback/quickpay.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply callback/quickpay.php:104]: Yes. You are right, that was an error case. I thought you were talking about "quickpay_create_payment_link" method. I fixed together with the bug from below "The following are not set: .... "

File: README.md [hunk 11]
```diff
@@ -1 +1,12 @@
...
+WHMCS integration with the QuickPay payment service provider.
+
+Based on existing integration but modified to use [payment links](https://learn.quickpay.net/tech-talk/payments/link/) instead of form
+
+Pull requests welcome!
+
+### Installation
+  * Upload the content to your modules/gateways/ folder
+  * Activate the gateway in the WHMCS admin area setup > payment > payment gateway
+  * Configure gateway
```
[the remark this answers is not in the document]
Comment (MichaelRotaru) [reply README.md:12]: "Please add install steps." All the installation steps are described. Is something extra we should add?

File: quickpay.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (cosminbosutar) [reply quickpay.php]: Understood

**Pull request merged**